linera-core/src/client/mod.rs

```diff
@@ -1395,11 +1395,63 @@ where
                 return Err(error.into());
             }
         };
-        for received_certificates in responses {
-            // Process received certificates.
-            self.receive_certificates_from_validator(received_certificates)
```

Comment:
With this change the `receive_certificates_from_validator` is used in one place only now. Would that other call-site also benefit from this parallelism? If so - why not move this logic to `receive_certificates_from_validator` instead?

Reply:
It wouldn't: The other call site is for the case where we receive a notification from _one_ validator about _one_ sender chain.

Reply:
but we still process all certificates separately, right? i.e. one call per certs while here we send all of them at once to a chain worker.

Reply:
Ah, they're still one-by-one but since they reuse worker.

Reply:
Yes, exactly. That's why I used `buffer_unordered` with (a bit less than) the number of chain workers, so that the tasks that run concurrently don't evict each other's workers.

Reply:
> It wouldn't: The other call site is for the case where we receive a notification from _one_ validator about _one_ sender chain.

In that case _this_ logic (the new one) is a generalisation of that case (I think) - i.e. it'd be as if the `certificates: BTreeMap` had only one key but the rest of the logic could remain the same.

Reply:
Not sure if this was clear from the previous comment – I'm still trying to check with you if it's possible to rewrite `receive_certificates_from_validator` in such a way that it can be used in both places and get the performance improvements you're making here.

Reply:
Yes, I understand. :+1: I'll look into it when I get back to working on this PR.

Reply:
Good idea, I deduplicated it.